### PR TITLE
`<spanstream>`: Use `static_cast` to `std::span` to avoid ill-formed constructs

### LINKED ISSUE #4879: `<spanstream>`: The `span` constructed by `basic_ispanstream`'s range constructor may be ill-formed
This range constructor is specified in [[ispanstream.cons]](https://eel.is/c++draft/ispanstream.cons) as:

> `template<class ROS> explicit basic_ispanstream(ROS&& s)`

> -3- _Constraints_: `ROS` models `ranges​::​borrowed_range`. `!convertible_to<ROS, std​::​span<charT>> && convertible_to<ROS, std​::​span<charT const>>` is `true`.

> -4- _Effects_: Let `sp` be `std​::​span<const charT>(std​::​forward<ROS>(s))`. Equivalent to:
>   `basic_ispanstream(std::span<charT>(const_cast<charT*>(sp.data()), sp.size()))`

Here, only `std​::​span<const charT>(std​::​forward<ROS>(s))` is required, but MSVC always constructs `span` via `std​::​span<const charT>(ranges:: data(s)), ranges::size(s))`, which is not always well-formed (for example, when `s` is not a `sized_range` or `contiguous_range`):

```cpp
#include <spanstream>

struct R {
  int* begin();
  std::unreachable_sentinel_t end();
  operator std::span<const int>() const;
};

int main() {
  R r{};
  std::basic_ispanstream<int> is{r}; // error
  is.span(r); // error
}
```
https://godbolt.org/z/Trv8d8qd3

The above example seems contrived, but libstdc++ accepts it as the standard guarantees it. (Whether this is worthy of LWG may be arguable).

### PR DESCRIPTION
**Pull request by euvre**
fix #4879

### CODE REVIEW

File: stl/inc/spanstream [hunk 1]
```diff
@@ -251,8 +252,8 @@ public:
         requires (
             !convertible_to<_ReadOnlyRange, _STD span<_Elem>> && convertible_to<_ReadOnlyRange, _STD span<const _Elem>>)
     void span(_ReadOnlyRange&& _Range) noexcept {
-        this->span(
-            _STD span<_Elem>{const_cast<_Elem*>(_RANGES data(_Range)), static_cast<size_t>(_RANGES size(_Range))});
+        this->span(_STD span<_Elem>{const_cast<_Elem*>(static_cast<_STD span<const _Elem>>(_Range).data()),
+            static_cast<size_t>(static_cast<_STD span<const _Elem>>(_Range).size())});
```
Comment (frederick-vs-ja) [inline stl/inc/spanstream:256]: I think we should construct a `span` variable once here.

File: stl/inc/spanstream [hunk 2]
```diff
@@ -218,7 +218,8 @@ public:
             !convertible_to<_ReadOnlyRange, _STD span<_Elem>> && convertible_to<_ReadOnlyRange, _STD span<const _Elem>>)
     explicit basic_ispanstream(_ReadOnlyRange&& _Range)
         : basic_ispanstream(
-            _STD span<_Elem>{const_cast<_Elem*>(_RANGES data(_Range)), static_cast<size_t>(_RANGES size(_Range))}) {}
+            _STD span<_Elem>{const_cast<_Elem*>(static_cast<_STD span<const _Elem>>(_Range).data()),
+                static_cast<size_t>(static_cast<_STD span<const _Elem>>(_Range).size())}) {}
```
Comment (frederick-vs-ja) [inline stl/inc/spanstream:222]: The return type `span<T, Ext>::size` is already `size_t`, so there shouldn't be `static_cast`.

**Review by frederick-vs-ja — CHANGES_REQUESTED** [review]

File: stl/inc/spanstream (hunk 2 above)
Comment (euvre) [reply stl/inc/spanstream]: I have fixed it.

**Comment by euvre** [comment]
@microsoft-github-policy-service agree

File: stl/inc/spanstream [hunk 3]
```diff
@@ -217,8 +217,8 @@ public:
         requires (
             !convertible_to<_ReadOnlyRange, _STD span<_Elem>> && convertible_to<_ReadOnlyRange, _STD span<const _Elem>>)
     explicit basic_ispanstream(_ReadOnlyRange&& _Range)
-        : basic_ispanstream(
-            _STD span<_Elem>{const_cast<_Elem*>(_RANGES data(_Range)), static_cast<size_t>(_RANGES size(_Range))}) {}
+        : basic_ispanstream(_STD span<_Elem>{const_cast<_Elem*>(static_cast<_STD span<const _Elem>>(_Range).data()),
+            static_cast<_STD span<const _Elem>>(_Range).size()}) {}
```
Comment (CaseyCarter) [inline stl/inc/spanstream:221]: This converts the argument twice, and without using `forward` to restore the value category, both of which are user-observable and not the intent of https://eel.is/c++draft/ispanstream.cons#4. I think we need another `explicit` non-template constructor taking `span<const _Elem>` for this constructor template to delegate to. Something like:
```suggestion
        : basic_ispanstream(static_cast<_STD span<const _Elem>>(_STD forward<_ReadOnlyRange>(_Range))) {}

    explicit basic_ispanstream(_STD span<const _Elem> _Span) : basic_ispanstream(_STD span<_Elem>{const_cast<_Elem*>(_Span.data()), _Span.size()}) {}
```
(Evidence that adding this constructor won't introduce ambiguity: https://www.godbolt.org/z/6KT7fqKof.)

File: tests/std/tests/GH_004879_span_constructed_by_basic_ispanstream/test.compile.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// GH-4879 <spanstream>: The span constructed by basic_ispanstream's range constructor may be ill-formed
```
Comment (CaseyCarter) [inline tests/std/tests/GH_004879_span_constructed_by_basic_ispanstream/test.compile.pass.cpp:4]: This seems pretty tiny to justify a separate test. Could we just add it to `tests\std\tests\P0448R4_spanstream\test.cpp`?

File: tests/std/tests/GH_004879_span_constructed_by_basic_ispanstream/test.compile.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+#include <spanstream>
+
+struct R {
+    char* begin();
+    std::unreachable_sentinel_t end();
+    operator std::span<const char>();
+};
+
+int main() {
```
Comment (CaseyCarter) [inline tests/std/tests/GH_004879_span_constructed_by_basic_ispanstream/test.compile.pass.cpp:14]: Our compile-only tests traditionally don't have a `main` as a reminder that they never run. We should rename this function to something like `void test()`. (You can pick a less boring name 😸.)

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: tests/std/tests/GH_004879_span_constructed_by_basic_ispanstream/test.compile.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// GH-4879 <spanstream>: The span constructed by basic_ispanstream's range constructor may be ill-formed
+
+#include <spanstream>
+
+struct R {
+    char* begin();
+    std::unreachable_sentinel_t end();
```
Comment (CaseyCarter) [inline tests/std/tests/GH_004879_span_constructed_by_basic_ispanstream/test.compile.pass.cpp:10]: Missing `#include <iterator>`.

File: tests/std/tests/GH_004879_span_constructed_by_basic_ispanstream/test.compile.pass.cpp [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// GH-4879 <spanstream>: The span constructed by basic_ispanstream's range constructor may be ill-formed
+
+#include <spanstream>
+
+struct R {
+    char* begin();
+    std::unreachable_sentinel_t end();
+    operator std::span<const char>();
```
Comment (CaseyCarter) [inline tests/std/tests/GH_004879_span_constructed_by_basic_ispanstream/test.compile.pass.cpp:11]: `#include <span>` wouldn't hurt, although it's admittedly hard to imagine how `<spanstream>` could work without at least a forward declaration of `span`.

Comment (euvre) [reply tests/std/tests/GH_004879_span_constructed_by_basic_ispanstream/test.compile.pass.cpp]: Thanks for your suggestions for improvement. I have fixed those defects.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks, looks good! I pushed changes for the issues I found.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this bug in a new feature and thoroughly testing it! :lady_beetle: :hammer_and_wrench: :white_check_mark:

**Pull request merged**